File: ts/lang/result.ts
/**
 * Copyright (c) 2023 The OtterOP Authors. All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 *    * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *    * Neither the name of Confluent Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 */


export class Result<RES, ERR> {

    private _res : RES;

    private _err : ERR;

    private constructor(_res : RES, _err : ERR) {
        this._res = _res;
        this._err = _err;
    }

    public isOK() : boolean {
        return this._err != null;
    }

    public err() : ERR {
        return this._err;
    }

    public unwrap() : RES {
        return this._res;
    }

    public static of<RES0, ERR0>(res : RES0, err : ERR0) : Result<RES0, ERR0> {
        return new Result<RES0,ERR0>(res, err);
    }
}

File: dotnet/Lang/Result.cs
namespace Otterop.Lang;

public class Result<RES, ERR>
{
    private RES _res;
    private ERR _err;
    private Result(RES _res, ERR _err)
    {
        this._res = _res;
        this._err = _err;
    }

    public bool IsOK()
    {
        return this._err != null;
    }

    public ERR Err()
    {
        return this._err;
    }

    public RES Unwrap()
    {
        return this._res;
    }

    public static Result<RES0, ERR0> Of<RES0, ERR0>(RES0 res, ERR0 err)
    {
        return new Result<RES0, ERR0>(res, err);
    }

}


public class Result
{
    public static Result<RES0, ERR0> Of<RES0, ERR0>(RES0 res, ERR0 err)
    {
        return Result<object, object>.Of<RES0, ERR0>(res, err);
    }

}
